src/hotspot/os/posix/os_posix.cpp

```diff
@@ -711,7 +711,16 @@ void* os::get_default_process_handle() {
 }
 
 void* os::dll_lookup(void* handle, const char* name) {
-  return dlsym(handle, name);
+  ::dlerror(); // Clear any previous error
```

Comment:
Uggghhh! It is awful that we need to do that. Now this needs to be guarded with `log_is_enabled(Debug, os)`. Otherwise we need to check for the performance hit of doing this, particularly during startup.

Reply:
The result are stored in per thread struct, so that method just load that thread local and check that result is NULL and returns. IMHO that's fine.

Reply:
Here is the data: Avg:187.913 ns Sum:40965 ns

Want me to add the guard? I don't mind.

Reply:
Sorry Robbin can you just re-explain why we need the initial call to `dl_error` here please. If the `dl_sym` is successful then we shouldn't be checking `dl_error`; and if not successful then `dl_error` should be returning that most recent failure.

Reply:
Hey, David, sure.

The dl error string is only clear by dlerror.
It can be set by dlopen/dlsym/dlclose on error (but never clear by these).
If someone used any of these before your call and failed, skipped calling dlerror,
there might be a pending error string.

dlsym returns NULL on ERROR, but the value of a symbol may also be NULL.

This means we need to use both dlerror return value plus dlsym return value to figure out what was going on.
dlsym = NULL, dlerror = NULL   => symbol was NULL
dlsym = NULL, dlerror = "error" => real error
dlsym = pointer, dlerror = ignored = found usable symbol

"Since the value of the symbol could actually be NULL (so that a NULL return from dlsym() need not indicate an error), the correct way to test for an error is to call dlerror() to clear any old error conditions, then call dlsym(), and then call dlerror() again, saving its return value into a variable, and check whether this saved value is not NULL."

Reply:
Make sense?

Reply:
Not from a programming perspective! That would be like successful syscalls not resetting errno to zero. Then again maybe they don't? Hmmm.

If this returns the last error that ever happened then all uses of the potential error producing functions should always be followed by a call to `dlerror` if an error occurred. I'm guessing we don't actually do that? So the one place we do check has to make a dummy call to clear the error first - blegghh! :(

Reply:
Agreed